### PR TITLE
2022-04-01 MSF Phase 2 - General refactoring BigCommerce

### PR DESCRIPTION
**Pull request by mrassudishkin**
[VT-5512] : https://github.com/skuvault-integrations/bigCommerceAccess/pull/25
[VT-5514] : https://github.com/skuvault-integrations/bigCommerceAccess/pull/28
[VT-5513] : https://github.com/skuvault-integrations/bigCommerceAccess/pull/30
[VT-5524] : https://github.com/skuvault-integrations/bigCommerceAccess/pull/31

This PR is comon PR for phase2 and review core team.

- Updated pagination with help meta information and added some tests.
- Removed ActionPolices and installed Polly.
- Extracted API services to abstract the logic for getting data from BC API;
- Moved ILogger and WebRequestServices out of the domain services and inject them;
- Removed logic of PageAdjuster
- Removed variants from request GetProducts
- Added separate request for variants. This request works by ids.
- Add unit tests for channels, orders and products

[VT-5512]: https://agileharbor.atlassian.net/browse/VT-5512?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[VT-5514]: https://agileharbor.atlassian.net/browse/VT-5514?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[VT-5513]: https://agileharbor.atlassian.net/browse/VT-5513?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[VT-5524]: https://agileharbor.atlassian.net/browse/VT-5524?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by dk-skuvault — APPROVED** [review]

File: src/BigCommerceAccessTests/IntegrationTests/ChannelsIntegrationTests.cs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+using System.Threading;
+using System.Threading.Tasks;
+using BigCommerceAccess.Models.Channel;
+using FluentAssertions;
+using NUnit.Framework;
+
+namespace BigCommerceAccessTests.IntegrationTests
+{
+	[ SuppressMessage( "Naming", "ASYNC0001:Asynchronous method names should end with Async" ) ]
+	[ SuppressMessage( "Performance", "ASYNC0004:Use ConfigureAwait(false) on await expression" ) ]
```
Comment (slav) [inline src/BigCommerceAccessTests/IntegrationTests/ChannelsIntegrationTests.cs:13]: Why not just add this at tests project csproj level?

File: src/BigCommerceAccessTests/Misc/PaginationTests.cs [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+		private List< string > _data = new List< string >();
+		private const int PageSize = 5;
+
+		[ SetUp ]
+		public void Init()
+		{
+		}
+
+		[ Test ]
+		public void GetDataPagesAsync_WhenNumberOfItemsIsLesserThanPageSize()
```
Comment (slav) [inline src/BigCommerceAccessTests/Misc/PaginationTests.cs:25]: Did you mean to say Smaller? I'm not sure if Lesser is a real word.

File: src/BigCommerceAccessTests/Misc/PaginationTests.cs [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+		{
+			// Arrange
+			this._data = Enumerable.Range( 1, PageSize - 1 ).Select( x => x.ToString() ).ToList();
+
+			// Act
+			var data = PaginationUtil.GetDataByPagesAsync(
```
Comment (slav) [inline src/BigCommerceAccessTests/Misc/PaginationTests.cs:31]: Why not make this test async and then await for the data?

File: src/BigCommerceAccessTests/Misc/RetryPolicyTests.cs [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+						Guid.NewGuid().ToString(),
+						BigCommerceCommand.GetProductsChannelAssignments.Command )
+					.ConfigureAwait( false )
+			);
+
+			// Assert
+			executionCount.Should().Be( maxRetryAttempts + 1 );
+
+			logger.Received( 0 ).LogCriticalException( exception, Arg.Any< string >(), Arg.Any< string >(), Arg.Any< string >() );
+			logger.Received( maxRetryAttempts ).LogRetryException(
```
Comment (slav) [inline src/BigCommerceAccessTests/Misc/RetryPolicyTests.cs:82]: Are the parameters really different from one run to the next? We should try to provide specific expected parmaeters when we can, and use Any if we it changes from run to run and we cannot determine what it is. Otherwise we're missing part of the test, don't we?

File: src/BigCommerceAccessTests/Misc/RetryPolicyTests.cs [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+		public async Task ExecuteAsync_RetriesToTheLimit_WhenSpecificStatusCodeInResponse( int httpStatusCode )
+		{
+			// Arrange
+			var maxRetryAttempts = 2;
+			var retryPolicy = new RetryPolicy( this._defaultLogger, maxRetryAttempts, DelayBaseInMs, DelayIncrementInMs );
+			var executionCount = 0;
+
+			// Act
+			await retryPolicy
+				.ExecuteAsync( async () => {
```
Comment (slav) [inline src/BigCommerceAccessTests/Misc/RetryPolicyTests.cs:147]: Does the part inside ExecuteAsync the same frome one test to the next? Should we refactor it to a common method we call?

File: src/BigCommerceAccessTests/UnitTests/OrdersUnitTests.cs [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+{
+	internal sealed class OrdersUnitTests: UnitTestsBase
+	{
+		private IOrdersApiService _ordersApiService;
+
+		private BigCommerceOrdersV2Service _bigCommerceOrdersV2Service;
+		private BigCommerceOrdersV3Service _bigCommerceOrdersV3Service;
+
+		[ SetUp ]
+		public new void Init()
```
Comment (slav) [inline src/BigCommerceAccessTests/UnitTests/OrdersUnitTests.cs:27]: This is very complicated setup and hard to follow. There're several options:
* Pull out to separate methods, each with good name
* Leave inline // Comment describing what each setup does

Both approaches aim to make it clear (document) what you intend the code to do.

File: src/BigCommerceAccessTests/UnitTests/OrdersUnitTests.cs [hunk 7]
```diff
@@ -0,0 +1,334 @@
...
+						new BigCommerceLimits(),
+						this.GetMetaData(
+							1,
+							1,
+							orderShippingAddresses.Count ),
+						HttpStatusCode.OK ) ) );
+		}
+
+		[ Test ]
+		public async Task GetOrdersAsync_ReturnsAllOrders_WhenWeUseV2Service()
```
Comment (slav) [inline src/BigCommerceAccessTests/UnitTests/OrdersUnitTests.cs:168]: Looking at this I've no idea why it works this way, and if it breaks, why it broke or how to fix. Why are we getting received 1, 1, then 2, 2? Is it because it was setup that way?

File: src/BigCommerceAccessTests/UnitTests/ProductsUnitTests.cs [hunk 8]
```diff
@@ -0,0 +1,623 @@
...
+			{
+				new BigCommerceProductChannelAssignment { ProductId = 1, ChannelId = 1 },
+				new BigCommerceProductChannelAssignment { ProductId = 2, ChannelId = 1 },
+				new BigCommerceProductChannelAssignment { ProductId = 3, ChannelId = 1 },
+				new BigCommerceProductChannelAssignment { ProductId = 4, ChannelId = 2 },
+				new BigCommerceProductChannelAssignment { ProductId = 5, ChannelId = 2 }
+			};
+
+		[ SetUp ]
+		public new void Init()
```
Comment (slav) [inline src/BigCommerceAccessTests/UnitTests/ProductsUnitTests.cs:35]: Again, can we document setup a little bit to make it clearer?

**Review by slav — CHANGES_REQUESTED** [review]
* Left inline comments
* A lot of empty whitespace at the end of lines formatting issues. Maybe reformat project?
* We'll need to switch to the other way to reference nuget packages (in csproj). Could do when switching projects to .net standard 2.0

File: src/BigCommerceAccessTests/Misc/PaginationTests.cs (hunk 2 above)
Comment (andrey-kozlov-skuvault) [reply src/BigCommerceAccessTests/Misc/PaginationTests.cs:25]: "Lesser" is a real word, but we should use "Less" here because we are talking about quantity here, not quality.
https://english.stackexchange.com/questions/24719/difference-between-less-and-lesser

File: src/BigCommerceAccess/Infrastructure/RetryPolicy.cs [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+		internal async Task< TResult > ExecuteAsync< TResult >(
+			Func< Task< TResult > > funcToExecute,
+			string marker,
+			string urlForLogging,
+			[ CallerMemberName ] string callerMethodName = null )
+			where TResult : IBigCommerceResponse
+		{
+			return await Policy< TResult >
+				.Handle< WebException >()
+				.OrResult( response => ( int )response.StatusCode >= 500
```
[the remark this answers is not in the document]
Comment (andrey-kozlov-skuvault) [reply src/BigCommerceAccess/Infrastructure/RetryPolicy.cs:51]: I think no, because here we are working with custom class `TResult : IBigCommerceResponse`, not with ``HttpResponseMessage` for which we have extensions: https://github.com/App-vNext/Polly.Extensions.Http/blob/master/src/Polly.Extensions.Http/HttpPolicyExtensions.cs#L43.
And we also extended transient errors with 429 status code, so there will be additional OrResult() anyway

**Review by andrey-kozlov-skuvault — APPROVED** [review]

File: src/BigCommerceAccess/ProductsService.cs [hunk 10]
```diff
@@ -15,22 +13,22 @@
 
 namespace BigCommerceAccess
 {
-	abstract class BigCommerceBaseProductsService : BigCommerceServiceBase
```
Comment (slav) [inline src/BigCommerceAccess/ProductsService.cs:15]: Didn't understand the class. Perhaps should be named then BigCommerceProductsServiceBase, similar to it's parent class?

File: src/BigCommerceAccess/Infrastructure/RetryPolicy.cs [hunk 11]
```diff
@@ -51,13 +49,12 @@ public RetryPolicy( ILogger logger, int retryAttempts = 10, int delayBaseInMs =
 				.OrResult( response => ( int )response.StatusCode >= 500 
 				                       || response.StatusCode == HttpStatusCode.RequestTimeout
 				                       || ( int )response.StatusCode == 429 )
-				.WaitAndRetryAsync( this.RetryAttempts,
-					retryAttempt => TimeSpan.FromMilliseconds( this.GetDelayBeforeNextAttempt( retryAttempt ) ),
+				.WaitAndRetryAsync( Backoff.ExponentialBackoff( TimeSpan.FromMilliseconds( DelayIncrementInMs ), this.RetryAttempts ),
```
Comment (slav) [inline src/BigCommerceAccess/Infrastructure/RetryPolicy.cs:52]: So what's the maximum delay we can get here? And what's total amount of time until we fail?

Comment (slav) [reply src/BigCommerceAccess/Infrastructure/RetryPolicy.cs:52]: Also, is it possible to provide cancellation token to backoff? Imagine we're trying to shut down cijobs, and here it's not allowing it to shut down because it keeps retrying.... Can we interrupt retries somehow?

**Review by slav — CHANGES_REQUESTED** [review]
Asked a couple of inline comments and to potentially add cancellation on retries, in case we're trying to shut down

File: src/BigCommerceAccess/Infrastructure/RetryPolicy.cs [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+using System.Runtime.CompilerServices;
+using System.Threading.Tasks;
+using BigCommerceAccess.Misc;
+using BigCommerceAccess.Models;
+using Polly;
+using Polly.Contrib.WaitAndRetry;
+
+namespace BigCommerceAccess.Infrastructure
+{
+	public sealed class RetryPolicy
```
Comment (chris-skuvault) [inline src/BigCommerceAccess/Infrastructure/RetryPolicy.cs:12]: Lets name this appropriate as its only used for Http related retries. Something like HttpRetryPolicy might make more sense.

**Review by chris-skuvault — CHANGES_REQUESTED** [review]

File: src/BigCommerceAccess/Infrastructure/RetryPolicy.cs [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+		/// <summary>
+		/// Executes function and retries on possible transient errors: if function fails with WebException, or returns status code more than 500, 408 (RequestTimeout) or 429(TooManyRequests)
+		/// </summary>
+		/// <param name="funcToExecute"></param>
+		/// <param name="marker"></param>
+		/// <param name="urlForLogging"></param>
+		/// <param name="callerMethodName"></param>
+		/// <typeparam name="TResult"></typeparam>
+		/// <returns></returns>
+		internal async Task< TResult > ExecuteAsync< TResult >(
```
[the remark this answers is not in the document]
Comment (slav) [reply src/BigCommerceAccess/Infrastructure/RetryPolicy.cs:40]: This and other comments will be irrelevant when we switch to ihttpclient factory with DI where polly rules are registered with the client factory in DI container.

File: src/BigCommerceAccess/Infrastructure/RetryPolicy.cs (hunk 11 above)
Comment (mrassudishkin) [reply src/BigCommerceAccess/Infrastructure/RetryPolicy.cs]: Added comment. 1000, 2000, 4000, 8000, 16000
I can't find way send CancellationToken to WaitAndRetryAsync.
